staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -23,57 +23,124 @@ import (
+	"k8s.io/apimachinery/pkg/types"
 	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/apis/audit"
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
 	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/filters"
 	authenticationapi "k8s.io/client-go/pkg/apis/authentication/v1"
 )
 
-var _ http.ResponseWriter = &auditResponseWriter{}
-
-type auditResponseWriter struct {
-	http.ResponseWriter
-	out io.Writer
-	id  string
+type AuditBackend interface {
+	// Log one event. After returning the passed event is reused. If the backend asynchronously process the
+	// event, it has to make a deep copy before. Per audit ID it might be that LogAuditEvent is called up to
+	// two times.
+	LogAuditEvent(ev *audit.Event) error
 }
 
-func (a *auditResponseWriter) WriteHeader(code int) {
-	line := fmt.Sprintf("%s AUDIT: id=%q response=\"%d\"\n", time.Now().Format(time.RFC3339Nano), a.id, code)
-	if _, err := fmt.Fprint(a.out, line); err != nil {
-		glog.Errorf("Unable to write audit log: %s, the error is: %v", line, err)
+func updateEventFromRequest(ev *audit.Event, req *http.Request, ctx request.Context) error {
+	attribs, err := GetAuthorizerAttributes(ctx)
+	if err != nil {
+		return err
 	}
 
-	a.ResponseWriter.WriteHeader(code)
-}
+	ev.Verb = attribs.GetVerb()
+	ev.RequestURI = req.URL.RequestURI()
+	ev.SourceIP = utilnet.GetClientIP(req).String()
 
-// fancyResponseWriterDelegator implements http.CloseNotifier, http.Flusher and
-// http.Hijacker which are needed to make certain http operation (e.g. watch, rsh, etc)
-// working.
-type fancyResponseWriterDelegator struct {
-	*auditResponseWriter
-}
+	if user := attribs.GetUser(); user != nil {
+		ev.User.Username = user.GetName()
+		ev.User.Extra = map[string]audit.ExtraValue{}
+		for k, v := range user.GetExtra() {
+			ev.User.Extra[k] = audit.ExtraValue(v)
+		}
+		ev.User.Groups = user.GetGroups()
+		ev.User.UID = user.GetUID()
+	}
 
-func (f *fancyResponseWriterDelegator) CloseNotify() <-chan bool {
-	return f.ResponseWriter.(http.CloseNotifier).CloseNotify()
-}
+	if asuser := req.Header.Get(authenticationapi.ImpersonateUserHeader); len(asuser) > 0 {
+		ev.Impersonate = &audit.UserInfo{
+			Username: asuser,
+		}
+		if requestedGroups := req.Header[authenticationapi.ImpersonateGroupHeader]; len(requestedGroups) > 0 {
+			ev.Impersonate.Groups = requestedGroups
+		}
 
-func (f *fancyResponseWriterDelegator) Flush() {
-	f.ResponseWriter.(http.Flusher).Flush()
+		ev.Impersonate.Extra = map[string]audit.ExtraValue{}
+		for k, v := range req.Header {
+			if !strings.HasPrefix(k, authenticationapi.ImpersonateUserExtraHeaderPrefix) {
+				continue
+			}
+			k = k[len(authenticationapi.ImpersonateUserExtraHeaderPrefix):]
+			ev.Impersonate.Extra[k] = audit.ExtraValue(v)
+		}
+	}
+
+	if attribs.IsResourceRequest() {
+		ev.ObjectRef = &audit.ObjectReference{
+			Kind:      attribs.GetResource(), // TODO: this is not the Kind, but the resource. RESTMapper?
```

Comment:
The resource is basically the type name in the http path. It matches with the Kind most of the time. I have added code to get the real Kind from the request body now.